From 1b53e804266f98b05af8168ee3b0b361b1a44250 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 11 Oct 2021 07:46:19 +0300
Subject: [PATCH] [SW-56687] gaudi2: add api for returning offset

Currently, the offset which is sent to LKD as part of response to
PREP_DESC message is set as descriptor base. The offset is
sent to LKD after applying mask of 0x3FFFFFF. This logic works for
all platforms expect for gaudi2 where incorrect offset is send to
LKD (for eg. In arc flow the offset sent is 0x1200000 instead of
0x200000).

This patch adds platform specific function to set offset.

Change-Id: I8ce719e7e02f7e0414bfc9659a598faba9d2521a
---

diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 43317d6..7fc5ead 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -976,3 +976,8 @@
 	regmap_writel(1, mmPCIE_WRAP_LBW_DRAIN_CFG);
 	regmap_writel(1, mmPCIE_WRAP_LBW_DRAIN_STAMP);
 }
+
+u32 lkd_dev_spec_get_desc_offset(uintptr_t base)
+{
+	return ((u32)base - SDRAM_BASE);
+}
diff --git a/lib/preboot/module_lkd.c b/lib/preboot/module_lkd.c
index c61d47f..7ddfc8e 100644
--- a/lib/preboot/module_lkd.c
+++ b/lib/preboot/module_lkd.c
@@ -504,6 +504,11 @@
 	return base;
 }
 
+__weak u32 lkd_dev_spec_get_desc_offset(uintptr_t base)
+{
+	return (u32)base;
+}
+
 static int module_lkd_prep_desc(int size, u32 *offset)
 {
 	struct mem_desc mdesc, img_desc;
@@ -545,7 +550,7 @@
 	cdesc_data_off = (u8 *)cdesc + sizeof(cdesc->header);
 	cdesc->header.crc32 = crc32(0, cdesc_data_off, cdesc->header.size);
 
-	*offset = (u32)mdesc.base;
+	*offset = lkd_dev_spec_get_desc_offset(mdesc.base);
 
 	preboot_security_dynamic_img_sram_open(MEM_LKD_PCI, ACCESS_RW);
 	if (sts.is_elbi_desc)
